2022-05-24 Future-based shuffle for Ray.md

```diff
@@ -0,0 +1,108 @@
+![image](https://user-images.githubusercontent.com/12445254/169472798-535b8229-8d82-473b-82a4-26136270a9c7.png)
+- Introduce shuffle operands implement burden because `get_data_keys` needs to be impelmented. 
+
+### Building shuffle DAG by reducer index
+Acutally, Mars' ChunkGraph has information about the upstream and downstream chunk dependencies. After converting to Ray DAG, this part of the information still exists. The reason for relying on data keys is that Mars replaced `ShuffleProxy` with `FetchShuffle`. The implementation of `Fetch/FetchShuffle` is an abstraction of P2P, which is actually inconsistent with the semantics of Ray DAG. Therefore, when using Ray Backend, a new set of Shuffle operators can be introduced to remove FetchShuffle and rely on Ray to automatically resolve subtask input arguments.
+
+#### Operand Changes
+- Add `n_reducers` and `reducer_ordinal` to all `MapReduceOperand` operands
+- Introduce a new group of `FetchShuffle` operands which only records `n_mappers` and `n_reducers` to reduce meta overhead:
+```python
+class FetchShuffleType(Enum):
+    FETCH_SHUFFLE_BY_INDEX = 0
+    FETCH_SHUFFLE_BY_DATA_KEY = 1
+
+class FetchShuffleByIndex(Operand):
+    _op_type_ = opcodes.FETCH_SHUFFLE_BY_INDEX
+    n_mappers = Int32Field("n_mappers")
+    n_reducers = Int32Field("n_reducers")
+
+class FetchShuffle(Operand):
+    _op_type_ = opcodes.FETCH_SHUFFLE_BY_DATA_KEY
+    source_keys = ListField("source_keys", FieldTypes.string)
+    source_idxes = ListField("source_idxes", FieldTypes.tuple(FieldTypes.uint64))
+    source_mappers = ListField("source_mappers", FieldTypes.uint16)
+```
+- Make shuffle operands mapper produce output with deterministic number and order.
+    - `TensorReshape` and `TensorBinCount` mapper needs change. 
+- _bagging.py uses `source_idxes`, which alsoe needs to be refactored.
+
+#### Graph Changes
+Add an option to replace `ShuffleProxy` to `FetchShuffleByIndex` when building subtask graph
+
+#### Using a ShuffleManager to manage Ray shuffle execution
+```python
+class ShuffleManager:
+    def __init__(self, subtask_graph):
+        self.subtask_graph = subtask_graph
+        # Build mapper index
+        # subtask: (shuffle_index, mapper_index)
+        # Build reducer index
+        # subtask: (shuffle_index, reducer_ordinal)
+    
+    def has_shuffle(self):
+        return self.num_shuffles > 0
+
+    def add_mapper_output_refs(
+        self, subtask, output_object_refs: List["ray.ObjectRef"]
+    ):
+        pass
+
+    def get_reducer_input_refs(self, subtask) -> List["ray.ObjectRef"]:
+        pass
+
+    def get_n_reducers(self, subtask):
+        pass
+```
+
+#### Build Shuffle DAG code
+Load reducer mapper inputs:
```

Comment:
More details needed showing how this implementation can coexist with Mars implementation.

Reply:
The `ShuffleManager` won't be shared by mars, but all operands change can be shared by all backends